File: BaekJoon/재귀/Z.py
"""
author: kyungmin
date: 23.12.14
title: Z
time: 30분
"""

import sys
input = sys.stdin.readline

N, r, c = map(int, input().rstrip().split())

cnt = 0
def z_search(n, x, y):
    global cnt

    if x == c and y == r:   
        return
    
    pn = 2**(n-1)

    # 1사분면
    if x <= c < x + pn\
    and y <= r < y + pn:
        z_search(n - 1, x, y)
        return
    cnt += 2**(2*n-2)

    # 2사분면
    if x + pn <= c\
    and y <= r < y + pn:
        z_search(n - 1, x + pn, y)
        return
    cnt += 2**(2*n-2)

    # 3사분면
    if x <= c < x + pn\
    and y + pn <= r:
        z_search(n - 1, x, y + pn)
        return
    cnt += 2**(2*n-2)

    # 4사분면
    if x + pn <= c\
    and y + pn <= r:
        z_search(n - 1, x + pn, y + pn)
        return
    cnt += 2**(2*n-2)

z_search(N, 0, 0)

print(int(cnt))

File: BaekjoonWithSwift/BaekjoonWithSwift/재귀/Z.swift
//
//  Z.swift
//  BaekjoonWithSwift
//
//  Created by kyungmin on 2023/12/15.
//

import Foundation

func Z() {
    let input = readLine()!.split(separator: " ").compactMap { Int($0) }
    let N = input[0]
    let r = input[1]
    let c = input[2]

    func zSearch(n: Int, r: Int, c: Int) -> Int {
        if n == 0 {
            return 0
        }
        // r이 y좌표 c이 x좌표
        if r < 1<<(n-1) && c < 1<<(n-1) { //1사분면
            return zSearch(n: n-1, r: r, c: c)
        } else if r < 1<<(n-1) && 1<<(n-1) <= c { //2사분면
            return (1<<(2*n-2)) + zSearch(n: n - 1, r: r, c: c - 1<<(n-1))
        } else if 1<<(n-1) <= r && c < 1<<(n-1) { //3사분면
            return (1<<(2*n-2)) * 2 + zSearch(n: n - 1, r: r - 1<<(n-1), c: c)
        } else { //4사분면
            return (1<<(2*n-2)) * 3 + zSearch(n: n - 1, r: r - 1<<(n-1), c: c - 1<<(n-1))
        }
    }
    print(zSearch(n: N, r: r, c: c))
}
